src/core/runner.ts

```diff
@@ -24,24 +39,75 @@ export async function run(rawOptions: RawOptions) {
+            );
+            return getResult();
+        }
+        if (
+            partialCheck.fences.length + partialCheck.sourceFiles.length >
+            options.partialCheckLimit
+        ) {
+            reportWarning(
+                `Skipping fence validation. The partial check had more than ${options.partialCheckLimit} changes`
+            );
+            return getResult();
+        }
+    }
+
     let sourceFileProvider: SourceFileProvider = options.looseRootFileDiscovery
         ? new FDirSourceFileProvider(options.project, options.rootDir)
         : new TypeScriptProgram(options.project);
 
-    // Do some sanity checks on the fences
-    validateTagsExist();
+    if (!partialCheck) {
+        // validating tags exist requires a full load of all fences
+        // we can't do this in partial check mode.
+        //
+        // Prefetching the full config set here avoids the overhead
+        // of partial fence loading, since we know we are loading
+        // the full fence set.
+        getConfigManager().getAllConfigs();
+        validateTagsExist();
+    } else {
+        reportWarning(
+            `skipping validateTagsExist. Cannot validate tag existence during partial checks`
+        );
+    }
+
+    if (partialCheck) {
+        // validate only those files specified on the command line,
+        // or included in the scope of changed fence files.
+        const fenceScopeFiles = await getSourceFilesNormalized(
+            sourceFileProvider,
+            partialCheck.fences.map((fencePath: NormalizedPath) => path.dirname(fencePath))
+        );
 
-    const normalizedFiles = await getSourceFilesNormalized(sourceFileProvider);
+        const normalizedFiles = [...partialCheck.sourceFiles, ...fenceScopeFiles];
 
-    // Limit the concurrent executed promises because
-    // otherwise we will open all the files at the same time and
-    // hit the MFILE error (when we hit rlimit)
-    await runWithConcurrentLimit(
-        options.maxConcurrentFenceJobs,
-        normalizedFiles,
-        (normalizedFile: NormalizedPath) => validateFile(normalizedFile, sourceFileProvider),
-        options.progress
-    );
+        // we have to limit the concurrent executed promises because
+        // otherwise we will open all the files at the same time and
+        // hit the MFILE error (when we hit rlimit)
+        await runWithConcurrentLimit(
+            options.maxConcurrentFenceJobs,
+            normalizedFiles,
+            (normalizedFile: NormalizedPath) => validateFile(normalizedFile, sourceFileProvider),
+            options.progress
+        );
+    } else {
+        const normalizedFiles = await getSourceFilesNormalized(sourceFileProvider);
 
+        // we have to limit the concurrent executed promises because
+        // otherwise we will open all the files at the same time and
+        // hit the MFILE error (when we hit rlimit)
+        await runWithConcurrentLimit(
```

Comment:
This call is exactly the same as in the above branch.  Can you refactor to calculate `normalizedFiles` (possibly in a separate function?) and then just call this once?  I think the flow will be clearer that way.